File: 4.Clients/WebApi/ClientApp/src/entities/process.ts
import { Stage } from './stage';
import { User } from './user';
import { Candidate } from './candidate';
import { ProcessStatusEnum } from './enums/process-status.enum';
import { SeniorityEnum } from './enums/seniority.enum';
import { HrStage } from './hr-stage';
import { TechnicalStage } from './technical-stage';
import { ClientStage } from './client-stage';
import { OfferStage } from './offer-stage';
import { EnglishLevelEnum } from './enums/english-level.enum';
import { ProcessCurrentStageEnum } from './enums/process-current-stage';
import { DeclineReason } from './declineReason';

export class Process {
    id: number;
    status: ProcessStatusEnum;
    currentStage: ProcessCurrentStageEnum;
    candidate: Candidate;
    userOwner: User;
    rejectionReason: string;
    declineReason: DeclineReason;
    actualSalary: number;
    wantedSalary: number;
    englishLevel: EnglishLevelEnum;
    seniority: SeniorityEnum;
    hrStage: HrStage;
    technicalStage: TechnicalStage;
    clientStage: ClientStage;
    offerStage: OfferStage;
    createdDate: Date;
    startDate?: Date;
    endDate?: Date;
    candidateId?: number;
    userDelegateId?: number;
    userOwnerId?: number;
    userDelegate?: User;
    declineReasonId?: number;
}

File: 1.Domain/Domain.Model/Process.cs
using Core;
using Domain.Model.Enum;
using System;

namespace Domain.Model
{
    public class Process : Entity<int>
    {
        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public ProcessStatus Status { get; set; }
        public ProcessCurrentStage CurrentStage { get; set; }
        public string RejectionReason { get; set; }
        public int? DeclineReasonId { get; set; }
        public DeclineReason DeclineReason { get; set; }
        public int? CandidateId { get; set; }
        public Candidate Candidate { get; set; }
        public int? UserOwnerId { get; set; }
        public User UserOwner { get; set; }
        public int? UserDelegateId { get; set; }
        public User UserDelegate { get; set; }
        public float ActualSalary { get { return HrStage.ActualSalary; } }
        public float WantedSalary { get { return HrStage.WantedSalary; } }        
        public EnglishLevel EnglishLevel { get { return HrStage.EnglishLevel; } }
        public Seniority Seniority { get {
                return (OfferStage.Status != StageStatus.NA ? OfferStage.Seniority : TechnicalStage.Seniority != 0 ? TechnicalStage.Seniority : TechnicalStage.AlternativeSeniority);
            }
        }
        public DateTime HireDate { get { return OfferStage.HireDate; } }
        public HrStage HrStage { get; set; }
        public TechnicalStage TechnicalStage { get; set; }
        public ClientStage ClientStage { get; set; }
        public OfferStage OfferStage { get; set; }
    }
}
